// @best/api-db/build/sql/postgres/db.js
"use strict";
/*
 * Copyright (c) 2019, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: MIT
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/MIT
 */
Object.defineProperty(exports, "__esModule", { value: true });
const pg_1 = require("pg");
const db_1 = require("../db");
const migrate_1 = require("./migrate");
const PGSSLMODE = process.env.PGSSLMODE;
class PostgresDatabase extends db_1.SQLDatabase {
    pool;
    migrated = false;
    constructor(config) {
        super();
        this.pool = new pg_1.Pool({
            connectionString: config.uri,
            ssl: config.ssl || PGSSLMODE ? { rejectUnauthorized: false } : false,
        });
    }
    query(text, params) {
        if (!this.migrated) {
            throw new Error('Database migrations have not been ensured.');
        }
        return this.pool.query(text, params);
    }
    async performMigrations() {
        await (0, migrate_1.migrate)(this.pool);
        this.migrated = true;
    }
}
exports.default = PostgresDatabase;
