// @best/utils/build/index.js
"use strict";
/*
 * Copyright (c) 2019, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: MIT
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/MIT
 */
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.coalesce = exports.normalizeSpecs = exports.normalizeClientConfig = exports.RunnerInterruption = exports.req = exports.matchSpecs = exports.proxifiedSocketOptions = exports.logError = exports.getSystemInfo = exports.cacheDirectory = exports.clearLine = exports.isCI = exports.isInteractive = exports.Logger = void 0;
var logger_1 = require("./logger");
Object.defineProperty(exports, "Logger", { enumerable: true, get: function () { return logger_1.Logger; } });
var is_interactive_1 = require("./is-interactive");
Object.defineProperty(exports, "isInteractive", { enumerable: true, get: function () { return is_interactive_1.isInteractive; } });
Object.defineProperty(exports, "isCI", { enumerable: true, get: function () { return is_interactive_1.isCI; } });
var clear_line_1 = require("./clear-line");
Object.defineProperty(exports, "clearLine", { enumerable: true, get: function () { return __importDefault(clear_line_1).default; } });
var cache_directory_1 = require("./cache-directory");
Object.defineProperty(exports, "cacheDirectory", { enumerable: true, get: function () { return __importDefault(cache_directory_1).default; } });
var system_info_1 = require("./system-info");
Object.defineProperty(exports, "getSystemInfo", { enumerable: true, get: function () { return system_info_1.getSystemInfo; } });
var log_error_1 = require("./log-error");
Object.defineProperty(exports, "logError", { enumerable: true, get: function () { return __importDefault(log_error_1).default; } });
var proxy_1 = require("./proxy");
Object.defineProperty(exports, "proxifiedSocketOptions", { enumerable: true, get: function () { return proxy_1.proxifiedSocketOptions; } });
var match_specs_1 = require("./match-specs");
Object.defineProperty(exports, "matchSpecs", { enumerable: true, get: function () { return match_specs_1.matchSpecs; } });
var req_1 = require("./req");
Object.defineProperty(exports, "req", { enumerable: true, get: function () { return req_1.req; } });
var runner_interruption_1 = require("./runner-interruption");
Object.defineProperty(exports, "RunnerInterruption", { enumerable: true, get: function () { return runner_interruption_1.RunnerInterruption; } });
var normalize_client_config_1 = require("./normalize-client-config");
Object.defineProperty(exports, "normalizeClientConfig", { enumerable: true, get: function () { return normalize_client_config_1.normalizeClientConfig; } });
Object.defineProperty(exports, "normalizeSpecs", { enumerable: true, get: function () { return normalize_client_config_1.normalizeSpecs; } });
var coalesce_1 = require("./coalesce");
Object.defineProperty(exports, "coalesce", { enumerable: true, get: function () { return coalesce_1.coalesce; } });
