// @best/utils/build/system-info.js
"use strict";
/*
 * Copyright (c) 2019, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: MIT
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/MIT
 */
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getSystemInfo = getSystemInfo;
const systeminformation_1 = __importDefault(require("systeminformation"));
async function getSystemInfo() {
    const system = await systeminformation_1.default.system();
    const cpu = await systeminformation_1.default.cpu();
    const { platform, distro, release, kernel, arch } = await systeminformation_1.default.osInfo();
    const { avgLoad } = await systeminformation_1.default.currentLoad();
    return {
        system: {
            manufacturer: system.manufacturer,
            model: system.model,
        },
        cpu: {
            manufacturer: cpu.manufacturer,
            brand: cpu.brand,
            family: cpu.family,
            model: cpu.model,
            speed: cpu.speed,
            cores: cpu.cores,
        },
        os: { platform, distro, release, kernel, arch },
        load: { cpuLoad: avgLoad },
    };
}
