// @best/builder/build/rollup-plugin-benchmark-import.js
"use strict";
/*
 * Copyright (c) 2019, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: MIT
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/MIT
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = default_1;
const PRIMITIVES = ['beforeAll', 'before', 'afterAll', 'after', 'benchmark', 'describe', 'run'];
function resolveModuleEntryFromPackage(module) {
    let modulePath;
    try {
        modulePath = require.resolve(`${module}/dist/index.js`);
    }
    catch (_e) {
        // intentional noop
    }
    return modulePath;
}
const BENCHMARK_RUNTIME_MODULE = '@best/runtime';
const BENCHMARK_IMPORT = `import { ${PRIMITIVES.join(',')} } from "${BENCHMARK_RUNTIME_MODULE}" \n`;
function default_1() {
    let input;
    return {
        name: 'benchmark-import',
        options(rollupOpts) {
            input = rollupOpts.input;
        },
        resolveId(id) {
            if (id === BENCHMARK_RUNTIME_MODULE) {
                return resolveModuleEntryFromPackage(BENCHMARK_RUNTIME_MODULE);
            }
            return undefined;
        },
        transform(src, id) {
            if (id === input) {
                src = BENCHMARK_IMPORT + src;
            }
            return { code: src, map: null };
        },
    };
}
