// @best/api-db/build/sql/transformer.js
"use strict";
/*
 * Copyright (c) 2019, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: MIT
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/MIT
 */
Object.defineProperty(exports, "__esModule", { value: true });
const denormalizeMetrics = (metrics) => {
    return Object.keys(metrics).map((key) => ({
        name: key,
        duration: metrics[key][0],
        stdDeviation: metrics[key][1],
    }));
};
exports.default = {
    projects: (query) => {
        return query.rows.map((row) => ({
            id: row.id,
            name: row.name,
            createdAt: row.created_at,
            lastReleaseDate: row.last_release_date,
        }));
    },
    snapshots: (query) => {
        return query.rows.map((row) => ({
            id: row.id,
            projectId: row.project_id,
            name: row.name,
            metrics: denormalizeMetrics(JSON.parse(row.metrics)),
            environmentHash: row.environment_hash,
            similarityHash: row.similarity_hash,
            commit: row.commit,
            commitDate: row.commit_date,
            temporary: row.temporary,
            createdAt: row.created_at,
            updatedAt: row.updated_at,
        }));
    },
};
